resharper/resharper-unity/src/CSharp/Feature/Services/QuickFixes/MoveQuickFixes/MonoBehaviourMoveUtil.cs

```diff
@@ -0,0 +1,263 @@
+            if (methodDeclaration == null)
+                return false;
+            
+            var statement = expression.GetContainingStatementLike();
+            if (statement == null)
+                return false;
+
+            var classDeclaration = expression.GetContainingNode<IClassDeclaration>();
+            if (classDeclaration == null)
+                return false;
+
+            return IsAvailableToMoveFromMethodToMethod(expression);
+        }
+
+        /// <summary>
+        /// Is node available to move outside the loop
+        /// </summary>
+        public static bool IsAvailableToMoveFromLoop([NotNull] ITreeNode toMove, [NotNull] ILoopStatement loop)
+        {
+            toMove.GetPsiServices().Locks.AssertReadAccessAllowed();
+
+            var sourceFile = toMove.GetSourceFile();
+            if (sourceFile == null)
+                return false;
+            
+            var loopStartOffset = loop.GetTreeStartOffset().Offset;
+            return IsAvailableToMoveInner(toMove, declaredElement => 
+                    declaredElement.GetDeclarationsIn(sourceFile)
+                    .FirstOrDefault(declaration => declaration.GetSourceFile() == sourceFile)?.GetTreeStartOffset().Offset < loopStartOffset
+            );
+        }
+
+        /// <summary>
+        /// Is node available to move from one method to another in MonoBehaviour class 
+        /// </summary>
+        public static bool IsAvailableToMoveFromMethodToMethod([NotNull] ITreeNode toMove)
+        {
+            toMove.GetPsiServices().Locks.AssertReadAccessAllowed();
+            
+            return IsAvailableToMoveInner(toMove);
+        }
+        
+        private static bool IsAvailableToMoveInner([NotNull] ITreeNode toMove, [CanBeNull] Func<IDeclaredElement, bool> isElementIgnored = null)
+        {
+            var nodeEnumerator = toMove.ThisAndDescendants();
+            while (nodeEnumerator.MoveNext())
+            {
+                var current = nodeEnumerator.Current;
+                IDeclaredElement declaredElement;
+                switch (current)
+                {
+                    case IDeclaration declaration:
+                        declaredElement = declaration.DeclaredElement;
+                        break;
+                    case IReferenceExpression referenceExpression:
+                        declaredElement = referenceExpression.Reference.Resolve().DeclaredElement;
+                        break;
+                    default:
+                        declaredElement = null;
+                        break;
+                }
+                
+                if (declaredElement == null || isElementIgnored != null && isElementIgnored(declaredElement))
+                    continue;
+
+                // if declared element is local, we can't move our expression outside the method
+                switch (declaredElement)
+                {
+                    case ILocalVariableDeclaration _:
```

Comment:
Please do not check declared element types for declaration types.
While they sometimes share the actual implementation type this is just an implementation detail and should never be used.

In the current code e.g. `IParameterDeclaration` is unlikely to share the same implementation type as `IParameter`